guacamole/src/main/frontend/src/app/client/directives/guacClientNotification.js

```diff
@@ -0,0 +1,431 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * A directive for displaying a non-global notification describing the status
+ * of a specific Guacamole client, including prompts for any information
+ * necessary to continue the connection.
+ */
+angular.module('client').directive('guacClientNotification', [function guacClientNotification() {
+
+    var directive = {
+        restrict: 'E',
+        replace: true,
+        templateUrl: 'app/client/templates/guacClientNotification.html'
+    };
+
+    directive.scope = {
+
+        /**
+         * The client whose status should be displayed.
+         * 
+         * @type ManagedClient
+         */
+        client : '='
+        
+    };
+
+    directive.controller = ['$scope', '$injector', '$element',
+        function guacClientNotificationController($scope, $injector, $element) {
+   
+        // Required types
+        var ManagedClient      = $injector.get('ManagedClient');
+        var ManagedClientState = $injector.get('ManagedClientState');
+        var Protocol           = $injector.get('Protocol');
+
+        // Required services
+        var $location              = $injector.get('$location');
+        var authenticationService  = $injector.get('authenticationService');
+        var guacClientManager      = $injector.get('guacClientManager');
+        var requestService         = $injector.get('requestService');
+        var userPageService        = $injector.get('userPageService');
+
+        /**
+         * A Notification object describing the client status to display as a
+         * dialog or prompt, as would be accepted by guacNotification.showStatus(),
+         * or false if no status should be shown.
+         *
+         * @type {Notification|Object|Boolean}
+         */
+        $scope.status = false;
+
+        /**
+         * All client error codes handled and passed off for translation. Any error
+         * code not present in this list will be represented by the "DEFAULT"
+         * translation.
+         */
+        var CLIENT_ERRORS = {
```

Comment:
Have you considered the `Set` type? It's a lot less awkward than this.

Reply:
I haven't, no. Is that something that will get magically compiled down to essentially this? I know things like the arrow notation, `const`, etc. are compiled away during the build, allowing us to maintain IE10+ compatibility, but I'm not sure about full-fledged objects/types like `Set`.

Reply:
I'll just do a quick test refactoring this one instance to `Set` and see what code is emitted.

Reply:
No dice:

![Screenshot 2021-07-07 9 38 21 PM](https://user-images.githubusercontent.com/4632905/124863434-ea47c980-df6b-11eb-8cb0-07499231843a.png)

If we wanted to use `Set`, we'd need a polyfill.

Reply:
OK - migrated all directly touched code from `var` to `const`/`let` via rebase. I avoided making sweeping changes, updating only the code that is already within this diff.